lib/pure/uri.nim

```diff
@@ -153,6 +163,41 @@ func encodeQuery*(query: openArray[(string, string)], usePlus = true,
       result.add('=')
       result.add(encodeUrl(val, usePlus))
 
+iterator decodeQuery*(data: string): tuple[key, value: TaintedString] =
+  ## Reads and decodes query string ``data`` and yields the (name, value) pairs the
+  ## data consists of.
+  runnableExamples:
+    var queryData: seq[(string, string)]
+    for (key, value) in decodeQuery("foo=1&bar=2"):
+      queryData.add((key, value))
+    doAssert queryData == @[("foo", "1"), ("bar", "2")]
+
+  proc parseData(data: string, i: int, field: var string): int =
+    result = i
+    while result < data.len:
+      case data[result]
+      of '%': add(field, decodePercent(data, result))
+      of '+': add(field, ' ')
+      of '=', '&': break
+      else: add(field, data[result])
+      inc(result)
+
+  var i = 0
+  var name = ""
+  var value = ""
+  # decode everything in one pass:
+  while i < data.len:
+    setLen(name, 0) # reuse memory
+    i = parseData(data, i, name)
+    setLen(value, 0) # reuse memory
+    if i < data.len and data[i] == '=':
+      inc(i) # skip '='
+      i = parseData(data, i, value)
+    yield (name.TaintedString, value.TaintedString)
+    if i < data.len:
+      if data[i] == '&': inc(i)
+      else: cgiError("'&' expected")
```

Comment:
* inline the code: `else: raise newException(CgiError, "'&' expected")` and remove `cgiError`
* IMO CgiError should be `ValueError` but it's tricky if it's used by cgi.nim; maybe it can all work though, by making cgi.nim use: `type CgiError = ValueError` ?

Reply:
cgi.nim could also just catch ValueError and reraise CgiError.

Reply:
I think the `reraise` idea is better.

Reply:
why though? what's the downside of the alias ?

Reply:
Original `CgiError` is object of `IOError`. It may break backwards compatibility a bit.

```nim
type
  CgiError* = object of IOError
  
  
try:
  raise newException(CgiError, "vgi")
except IOError:
  echo "True"
```

Reply:
reason more to deprecate cgi.decodeData (refs https://github.com/nim-lang/Nim/pull/15860#discussion_r518780631); if we deprecate `cgi.decodeData` I'm fine with this